os/drivers/video/video_null_driver.c

```diff
@@ -0,0 +1,1554 @@
+	FAR null_priv_t *priv = (FAR null_priv_t *) arg;
+
+	if (priv == NULL) {
+		return -EINVAL;
+	}
+
+	video_null_takesem(&priv->sem);
+	datasize = priv->trfsize;
+	/* Check if the transfer is already cancelled by applications */
+	if (priv->reqcancel) {
+		uvdbg("Transfer already cancelled!!\n");
+		video_null_givesem(&priv->sem);
+		return;
+	}
+	memcpy(priv->reqbuff, priv->trfbuff, datasize);
+	video_null_givesem(&priv->sem);
+
+	video_common_notify_dma_done(0, priv->reqtype, datasize, priv->priv_data);
+}
+
+/****************************************************************************
+ * Name: video_null_open
+ *
+ * Description:
+ *   This function is video dev ops which open and allocate all the resource
+ *   requried for video upper layer.
+ *
+ *   This function is called from the video driver using dev ops.
+ *
+ * Input Parameters:
+ *   video_private - Reference to private data.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+static int video_null_open(FAR void *video_private)
+{
+	int ret = OK;
+	uint16_t tmp;
+	struct video_lowerhalf_s *phalf = (struct video_lowerhalf_s *)video_private;
+	FAR null_priv_t *priv = NULL;
+
+	if (phalf == NULL) {
+		return -EINVAL;
+	}
+
+	priv = (FAR null_priv_t *) phalf->dev;
+
+	if (priv == NULL) {
+		return -EINVAL;
+	}
+
+	/* Increment the driver reference count */
+	video_null_takesem(&priv->sem);
+	tmp = priv->crefs + 1;
+	if (tmp == 0) {
+		/* More than 255 opens; uint8_t overflows to zero */
+
+		ret = -EMFILE;
+		goto errout;
+	}
+
+	/* Save the new open count */
+
+	priv->crefs = tmp;
+	video_null_givesem(&priv->sem);
+errout:
+	return ret;
+}
+
+/****************************************************************************
+ * Name: video_null_close
+ *
+ * Description:
+ *   This function is video dev ops which close and release all the resource
+ *   acquired by the video upper layer.
+ *
+ *   This function is called from the video driver using dev ops.
+ *
+ * Input Parameters:
+ *   video_private - Reference to private data.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+static int video_null_close(FAR void *video_private)
+{
+	int ret = OK;
+	struct video_lowerhalf_s *phalf = (struct video_lowerhalf_s *)video_private;
+	FAR null_priv_t *priv = NULL;
+
+	if (phalf == NULL) {
+		return -EINVAL;
+	}
+
+	priv = (FAR null_priv_t *) phalf->dev;
+
+	if (priv == NULL) {
+		return -EINVAL;
+	}
+
+	/* Decrement the driver reference count */
+	video_null_takesem(&priv->sem);
+	if (priv->crefs > 0) {
+		priv->crefs--;
+	}
+	video_null_givesem(&priv->sem);
+	return ret;
+}
+
+/****************************************************************************
+ * Name: video_null_do_halfpush
+ *
+ * Description:
+ *   This function is video dev ops to send the partial data.
```

Comment:
According to description here, it need to send partial data to application, am I right?
Then some more logic is needed...

Reply:
Yes right, we need to send a partial data reading from filesystem, currently that activity on progress. we will make that as a separate patch

Reply:
pipeline buffer & mq or callback will be better than write on fs.

Reply:
We are still not clear on halfpush calls. Currently we are stabilizing the basic code.
Once we are sure on halfpush API functionality, we will make changes accordingly.
Since this the initial patch for null-driver, kindly approve the same.
Parallely we will enhance the features in upcoming patches.